File: bifrost-py/src/bifrost/server/method.py
"""RPC method primitive — port of ``bifrost-ts/src/server/method.ts``.

Python handler signature differs from TS: instead of binding
``this = ClientNode`` the handler receives ``(node, params)``
explicitly. Everything else — schema validation, middleware chain,
cache-key computation via ``EJSON.stringify``, ambient context via
``BifrostContext.run`` — matches the TS runtime behavior.
"""

from __future__ import annotations

import inspect
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from bifrost.ejson.parse_stringify import stringify
from bifrost.ejson.presentation import Presentation
from bifrost.protocol.constants import ServerEvents
from bifrost.server.context import BifrostContext
from bifrost.server.schema import SchemaValidator
from bifrost.utils.errors import Errors, SchemaValidationError

if TYPE_CHECKING:
    from bifrost.server.client_node import ClientNode
    from bifrost.server.server import Server


MethodFunction = Callable[..., Any]
Middleware = Callable[..., Any]


DEFAULT_MAX_AGE_MS = 60_000


@dataclass(slots=True)
class MethodOptions:
    cache: bool = False
    max_age: int = DEFAULT_MAX_AGE_MS
    protected: bool = False
    middleware: list[Middleware] = field(default_factory=list)
    schema: SchemaValidator | None = None


async def _call_possibly_async(fn: Callable[..., Any], *args: Any) -> Any:
    result = fn(*args)
    if inspect.isawaitable(result):
        return await result
    return result


class Method:
    """Registered RPC method with validation, middleware, cache, and timing.

    Execution flow matches PROTOCOL.md sections 6.3 to 6.7:

    1. If a schema is attached, ``safe_parse`` the params and throw
       :class:`SchemaValidationError` on failure. Params normalize to
       ``{}`` when None.
    2. Enter :func:`BifrostContext.run` so handlers see the ambient
       execution context.
    3. Run the middleware chain, each step receiving ``(node, params)``
       and returning transformed params.
    4. Invoke the handler with ``(node, final_params)``.
    5. Emit :class:`ServerEvents.METHOD_EXECUTION` with timing.
    """

    uuid: str
    name: str
    server: Server
    fn: MethodFunction
    is_protected: bool
    middleware: list[Middleware]
    schema: SchemaValidator | None

    def __init__(
        self,
        server: Server,
        name: str,
        fn: MethodFunction,
        opts: MethodOptions | None = None,
    ) -> None:
        opts = opts or MethodOptions()
        self.uuid = Presentation.uuid()
        self.name = name
        self.server = server
        self.is_protected = bool(opts.protected)
        self.middleware = list(opts.middleware)
        self.schema = opts.schema

        self.fn = (
            _memoize(fn, max_age_ms=opts.max_age) if opts.cache else fn
        )

    async def run_middleware(
        self, params: Any, node: ClientNode | None
    ) -> Any:
        """Pipe params through middleware; each callable receives ``(node, params)``."""

        if not self.middleware:
            return params
        buffer = params
        for step in self.middleware:
            buffer = await _call_possibly_async(step, node, buffer)
        return buffer

    async def exec(self, params: Any, node: ClientNode | None = None) -> Any:
        """Execute the method. See class docstring for the flow."""

        start = time.perf_counter()

        clean_params = params

        if self.schema is not None:
            to_validate = params if params is not None else {}
            result = self.schema.safe_parse(to_validate)
            if not result.success:
                error_messages = [issue.format() for issue in result.issues]
                raise SchemaValidationError(
                    f"{Errors.INVALID_PARAMS.value}: {', '.join(error_messages)}",
                    error_messages,
                )
            clean_params = result.data

        caller_context = node.context if node is not None else None
        with BifrostContext.run(Presentation.uuid(), caller_context):
            final_params = await self.run_middleware(clean_params, node)
            returned = await _call_possibly_async(self.fn, node, final_params)

        elapsed_ms = (time.perf_counter() - start) * 1000

        self.server.emit_server_event(
            ServerEvents.METHOD_EXECUTION,
            {
                "method": self.name,
                "time": elapsed_ms,
                "params": clean_params,
                "result": returned,
            },
        )

        return returned


def _memoize(
    fn: MethodFunction, *, max_age_ms: int
) -> MethodFunction:
    """Cache-key algorithm MUST match TS: ``EJSON.stringify(first_arg)``
    with non-canonical output (PROTOCOL.md §6.3).

    The Python handler receives ``(node, params)``; the cache key is
    computed from ``params`` (the first RPC argument) so it aligns with
    the TS ``args[0]`` convention.
    """

    cache: dict[str, tuple[Any, float]] = {}

    async def memoized(node: ClientNode | None, params: Any) -> Any:
        key = stringify(params)
        now_ms = time.monotonic() * 1000
        hit = cache.get(key)
        if hit is not None and (now_ms - hit[1]) < max_age_ms:
            return hit[0]
        result = await _call_possibly_async(fn, node, params)
        cache[key] = (result, now_ms)
        return result

    return memoized

File: bifrost-rs/crates/bifrost-runtime/src/method.rs
//! RPC method primitive — port of `bifrost-py/src/bifrost/server/method.py`.

use crate::client_node::ClientNode;
use crate::context::BifrostContext;
use crate::error::{BifrostError, SchemaValidationError};
use crate::schema::SchemaValidatorArc;
use bifrost_ejson::stable_stringify;
use bifrost_protocol::errors::INVALID_PARAMS;
use futures::future::BoxFuture;
use serde_json::{Value, json};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};
use uuid::Uuid;

pub type BoxResult = Result<Value, BifrostError>;
pub type RpcFuture = BoxFuture<'static, BoxResult>;
pub type RpcHandler =
    Arc<dyn Fn(Arc<ClientNode>, Value) -> RpcFuture + Send + Sync>;
pub type Middleware =
    Arc<dyn Fn(Arc<ClientNode>, Value) -> RpcFuture + Send + Sync>;

pub const DEFAULT_MAX_AGE_MS: u64 = 60_000;

#[derive(Clone, Default)]
pub struct MethodOptions {
    pub cache: bool,
    pub max_age_ms: Option<u64>,
    pub protected: bool,
    pub middleware: Vec<Middleware>,
    pub schema: Option<SchemaValidatorArc>,
}

pub struct Method {
    pub uuid: String,
    pub name: String,
    pub is_protected: bool,
    handler: RpcHandler,
    middleware: Vec<Middleware>,
    schema: Option<SchemaValidatorArc>,
    cache: Option<Arc<MethodCache>>,
}

struct MethodCache {
    max_age: Duration,
    entries: RwLock<HashMap<String, (Value, Instant)>>,
}

impl Method {
    pub fn new(name: impl Into<String>, handler: RpcHandler, opts: MethodOptions) -> Arc<Self> {
        let cache = if opts.cache {
            Some(Arc::new(MethodCache {
                max_age: Duration::from_millis(opts.max_age_ms.unwrap_or(DEFAULT_MAX_AGE_MS)),
                entries: RwLock::new(HashMap::new()),
            }))
        } else {
            None
        };
        Arc::new(Self {
            uuid: Uuid::new_v4().to_string(),
            name: name.into(),
            is_protected: opts.protected,
            handler,
            middleware: opts.middleware,
            schema: opts.schema,
            cache,
        })
    }

    pub async fn exec(&self, params: Value, node: Arc<ClientNode>) -> BoxResult {
        // 1. Schema validation.
        let mut clean = params;
        if let Some(schema) = &self.schema {
            let to_validate = if clean.is_null() { json!({}) } else { clean.clone() };
            let result = schema.safe_parse(&to_validate);
            if !result.success {
                let messages: Vec<String> = result.issues.iter().map(|i| i.format()).collect();
                let message = format!("{INVALID_PARAMS}: {}", messages.join(", "));
                return Err(BifrostError::Schema(SchemaValidationError {
                    message,
                    errors: messages,
                }));
            }
            clean = result.data.unwrap_or(Value::Null);
        }

        // 2. Cache lookup.
        if let Some(cache) = &self.cache {
            let key = stable_stringify::stringify(&clean);
            if let Some(hit) = cache_get(cache, &key) {
                return Ok(hit);
            }
            // Cache miss — fall through to execute and remember.
            let result = run_with_context(self, clean.clone(), node).await?;
            cache_put(cache, key, result.clone());
            return Ok(result);
        }

        run_with_context(self, clean, node).await
    }
}

async fn run_with_context(
    method: &Method,
    clean: Value,
    node: Arc<ClientNode>,
) -> BoxResult {
    let ctx = node.context.read().expect("ClientNode.context poisoned").clone();
    BifrostContext::run(Uuid::new_v4().to_string(), ctx, async move {
        let mut buffer = clean;
        for step in &method.middleware {
            buffer = step(node.clone(), buffer).await?;
        }
        (method.handler)(node, buffer).await
    })
    .await
}

fn cache_get(cache: &MethodCache, key: &str) -> Option<Value> {
    let entries = cache.entries.read().expect("cache poisoned");
    let hit = entries.get(key)?;
    if hit.1.elapsed() < cache.max_age {
        Some(hit.0.clone())
    } else {
        None
    }
}

fn cache_put(cache: &MethodCache, key: String, value: Value) {
    let mut entries = cache.entries.write().expect("cache poisoned");
    entries.insert(key, (value, Instant::now()));
}
